Migrate to RxJava 3

MrAsterisco opened the pull request:
This PR introduces the following update: support for RxJava 3 (potentially closing #365 ). 

I wanted to use ReactiveNetwork in a new project that targets RxJava 3, so I figured I might as well make the transition myself.

A few major things have changed:
- I have changed the package name to "rx3".
- The Build Tools are now on "29.0.2" to support the new Gradle plugin in Android Studio 4.
- Gradle version is now at "6.3".
- The `onError` method of the `NetworkObservingStrategy` interface is now using a `Throwable` instead of a simple `Exception`. This had to be changed because a lot of methods inside the various implementations of the interface can throw a generic `Throwable`.
- The `android.enableUnitTestBinaryResources` has been removed because it is no longer supported.

The rest of the changes are just related to the new version of RxJava (mainly changing the package references).

All unit tests pass, all code styles checks pass. Both example apps work perfectly (I had to add the Java compatibility option to the Kotlin app to make it build). The [build on Travis](https://travis-ci.org/github/MrAsterisco/ReactiveNetwork) works.

Please, let me know if you're interested in merging this PR and, if so, what would be the next steps.

MrAsterisco committed:
Migrate to RxJava 3

Fix code style

Update README and CHANGELOG

Apply code style, update README

Update artifact name

Update artifact name, proguard rules, version and version code

pwittchen commented:
Hi @MrAsterisco,

Thanks for this PR!

Could you please squash your commits into one and set target branch to `RxJava3.x`? I know it didn't exist before, but I just created it and I'd like to have changes related to RxJava3.x on that branch and `RxJava2.x` should contain code depending on RxJava2.x only.

Regards,
Piotr

pwittchen commented:
I already changed target branch (base) to `RxJava3.x`. I found out it was possible in this PR.

MrAsterisco commented:
Ok, great! Before squashing my commits, would you like me to go through the process of updating the Changelog file and generating the new documentation?

pwittchen commented:
You can update changelog and documentation in `README.md` file. I'll need to update JavaDocs and website on gh-pages later separately because that's another branch.

MrAsterisco commented:
Ok, done! I'm assuming that this will be a 4.0.0, since there is a (small) breaking change in the `NetworkObservingStrategy` interface.

pwittchen commented on README.md:
```diff
@@ -2,14 +2,15 @@
 [![Android Arsenal](https://img.shields.io/badge/Android%20Arsenal-ReactiveNetwork-brightgreen.svg?style=flat-square)](https://android-arsenal.com/details/1/2290)
 
-view website with documentation: [RxJava1.x](http://pwittchen.github.io/ReactiveNetwork/docs/RxJava1.x/), [**RxJava2.x**](http://pwittchen.github.io/ReactiveNetwork/docs/RxJava2.x/)
+view website with documentation: [RxJava1.x](http://pwittchen.github.io/ReactiveNetwork/docs/RxJava1.x/), [**RxJava2.x**](http://pwittchen.github.io/ReactiveNetwork/docs/RxJava2.x/), [RxJava3.x](http://pwittchen.github.io/ReactiveNetwork/docs/RxJava3.x/)
```
You can make RxJava3.x bold with markdown and remove bold font from RxJava2.x

pwittchen commented on README.md:
```diff
@@ -2,14 +2,15 @@
 |:--------------:|:-------:|:-----------:|:-------------:|:--------:|:-------------:|
 | | [`RxJava1.x`](https://github.com/pwittchen/ReactiveNetwork/tree/RxJava1.x) | `reactivenetwork` | [![Build Status for RxJava1.x](https://img.shields.io/travis/pwittchen/ReactiveNetwork/RxJava1.x.svg?style=flat-square)](https://travis-ci.org/pwittchen/ReactiveNetwork) | [![codecov](https://img.shields.io/codecov/c/github/pwittchen/ReactiveNetwork/RxJava1.x.svg?style=flat-square&label=coverage)] ...
 | :ballot_box_with_check: | [`RxJava2.x`](https://github.com/pwittchen/ReactiveNetwork/tree/RxJava2.x) | `reactivenetwork-rx2` | [![Build Status for RxJava2.x](https://img.shields.io/travis/pwittchen/ReactiveNetwork/RxJava2.x.svg?style=flat-square)](https://travis-ci.org/pwittchen/ReactiveNetwork) | [![codecov](https://img.shields.io/codecov/c/github/pwittchen/ReactiveNetwork/RxJava2.x.svg?style= ...
+| | [`RxJava3.x`](https://github.com/pwittchen/ReactiveNetwork/tree/RxJava3.x) | `reactivenetwork-rx3` | [![Build Status for RxJava3.x](https://img.shields.io/travis/pwittchen/ReactiveNetwork/RxJava3.x.svg?style=flat-square)](https://travis-ci.org/pwittchen/ReactiveNetwork) | [![codecov](https://img.shields.io/codecov/c/github/pwittchen/ReactiveNetwork/RxJava3.x.svg?style=flat-square&label=covera ...
```
You can add checkbox with markdown with `:ballot_box_with_check:` command for RxJava3.x and remove it from RxJava2.x

pwittchen commented on README.md:
```diff
@@ -600,7 +601,7 @@ Then commit and push your changes to `gh-pages` branch.
 --------------------------
 
-view website with documentation: [RxJava1.x](http://pwittchen.github.io/ReactiveNetwork/docs/RxJava1.x/), [**RxJava2.x**](http://pwittchen.github.io/ReactiveNetwork/docs/RxJava2.x/)
+view website with documentation: [RxJava1.x](http://pwittchen.github.io/ReactiveNetwork/docs/RxJava1.x/), [**RxJava2.x**](http://pwittchen.github.io/ReactiveNetwork/docs/RxJava2.x/), [RxJava3.x](http://pwittchen.github.io/ReactiveNetwork/docs/RxJava3.x/)
```
You can make RxJava3.x bold with markdown and remove bold font from RxJava2.x

pwittchen reviewed (changes requested):
Changes look fine. I have a few issues with the documentation in `README.md`. Moreover, I have a question: Did you use code formatter / code style described here: https://github.com/pwittchen/ReactiveNetwork#code-style ? If not, then please apply it and reformat code with it.

MrAsterisco commented:
OK! I've reformatted the code using the AndroidSquare style (about 30 files were not formatted correctly) and I've also updated the README as you asked (I've also changed the references to the RxJava2.x and the name of the artifact). 

I noticed there's a Contributors section, so I've added myself in there; hope that's ok.

pwittchen commented:
Looks good. There's one more change required for migration to the new RxJava version. In the `library/gradle.properties` file, we need to change:

```
POM_ARTIFACT_ID=reactivenetwork-rx2
```

to:

```
POM_ARTIFACT_ID=reactivenetwork-rx3
```

MrAsterisco commented:
Done! To be sure, I've also searched for "rx2" in the whole project and I've updated the proguard rules files to point to the correct package.

I've also updated `VERSION_NAME` and `VERSION_CODE` in `gradle.properties`.

pwittchen commented:
Merged. It will be available in the next release for RxJava3.x. Thanks for your contribution @MrAsterisco!
